File: tests/stress_test.py
#!/usr/bin/env python3

import threading
import time
import random
import sys
import argparse
from datetime import datetime
import psycopg2
from psycopg2.extras import execute_batch
import logging

class StressTest:
    def __init__(self, host='localhost', port=5432, database='postgres', user='postgres', password=''):
        self.host = host
        self.port = port
        self.database = database
        self.user = user
        self.password = password
        self.running = False
        self.errors = []
        self.stats = {
            'batch_inserts': 0,
            'point_inserts': 0,
            'updates': 0,
            'deletes': 0,
            'queries': 0,
            'consistency_checks': 0,
            'inconsistencies': 0
        }
        self.lock = threading.Lock()
        
    def get_connection(self):
        return psycopg2.connect(
            host=self.host,
            port=self.port,
            database=self.database,
            user=self.user,
            password=self.password,
            connect_timeout=10
        )
    
    def setup_tables(self):
        conn = self.get_connection()
        try:
            with conn.cursor() as cur:
                cur.execute("CREATE EXTENSION IF NOT EXISTS pg_mooncake;")
                
                cur.execute("DROP TABLE IF EXISTS stress_regular CASCADE;")
                cur.execute("DROP TABLE IF EXISTS stress_columnstore CASCADE;")
                
                cur.execute("""
                    CREATE TABLE stress_regular (
                        id SERIAL PRIMARY KEY,
                        value INTEGER,
                        text_data TEXT,
                        created_at TIMESTAMP DEFAULT NOW()
                    );
                """)
                
                cur.execute("CALL mooncake.create_table('stress_columnstore', 'stress_regular');")
                
            conn.commit()
            logging.info("Tables created successfully")
        except Exception as e:
            logging.error(f"Error setting up tables: {e}")
            raise
        finally:
            conn.close()
    
    def cleanup_tables(self):
        conn = self.get_connection()
        try:
            with conn.cursor() as cur:
                cur.execute("DROP TABLE IF EXISTS stress_regular CASCADE;")
                cur.execute("DROP TABLE IF EXISTS stress_columnstore CASCADE;")
            conn.commit()
            logging.info("Tables cleaned up successfully")
        except Exception as e:
            logging.error(f"Error cleaning up tables: {e}")
        finally:
            conn.close()
    
    def batch_insert_worker(self, batch_size=100, iterations=10):
        conn = self.get_connection()
        try:
            while self.running:
                data = [(random.randint(1, 10000), f'batch_text_{random.randint(1, 1000)}') 
                       for _ in range(batch_size)]
                
                with conn.cursor() as cur:
                    execute_batch(
                        cur,
                        "INSERT INTO stress_regular (value, text_data) VALUES (%s, %s)",
                        data
                    )
                conn.commit()
                
                with self.lock:
                    self.stats['batch_inserts'] += len(data)
                
                time.sleep(random.uniform(0.1, 0.5))
                
        except Exception as e:
            with self.lock:
                self.errors.append(f"Batch insert error: {e}")
        finally:
            conn.close()
    
    def point_operations_worker(self):
        conn = self.get_connection()
        try:
            while self.running:
                operation = random.choice(['insert', 'update', 'delete'])
                
                with conn.cursor() as cur:
                    if operation == 'insert':
                        cur.execute(
                            "INSERT INTO stress_regular (value, text_data) VALUES (%s, %s)",
                            (random.randint(1, 10000), f'point_text_{random.randint(1, 1000)}')
                        )
                        with self.lock:
                            self.stats['point_inserts'] += 1
                    
                    elif operation == 'update':
                        cur.execute(
                            "UPDATE stress_regular SET value = %s WHERE id = (SELECT id FROM stress_regular ORDER BY RANDOM() LIMIT 1)",
                            (random.randint(1, 10000),)
                        )
                        if cur.rowcount > 0:
                            with self.lock:
                                self.stats['updates'] += 1
                    
                    elif operation == 'delete':
                        cur.execute(
                            "DELETE FROM stress_regular WHERE id = (SELECT id FROM stress_regular ORDER BY RANDOM() LIMIT 1)"
                        )
                        if cur.rowcount > 0:
                            with self.lock:
                                self.stats['deletes'] += 1
                
                conn.commit()
                time.sleep(random.uniform(0.01, 0.1))
                
        except Exception as e:
            with self.lock:
                self.errors.append(f"Point operation error: {e}")
        finally:
            conn.close()
    
    def query_worker(self):
        conn = self.get_connection()
        try:
            while self.running:
                with conn.cursor() as cur:
                    query_type = random.choice(['count', 'aggregate', 'filter'])
                    
                    if query_type == 'count':
                        cur.execute("SELECT COUNT(*) FROM stress_columnstore")
                    elif query_type == 'aggregate':
                        cur.execute("SELECT AVG(value), MAX(value), MIN(value) FROM stress_columnstore")
                    elif query_type == 'filter':
                        cur.execute("SELECT * FROM stress_columnstore WHERE value > %s LIMIT 10", 
                                  (random.randint(1, 5000),))
                    
                    result = cur.fetchall()
                    
                    with self.lock:
                        self.stats['queries'] += 1
                
                time.sleep(random.uniform(0.05, 0.2))
                
        except Exception as e:
            with self.lock:
                self.errors.append(f"Query error: {e}")
        finally:
            conn.close()
    
    def consistency_checker(self):
        conn = self.get_connection()
        try:
            while self.running:
                with conn.cursor() as cur:
                    cur.execute("CALL mooncake.create_snapshot('stress_columnstore');")
                    
                    cur.execute("SELECT COUNT(*) FROM stress_regular")
                    regular_count = cur.fetchone()[0]
                    
                    cur.execute("SELECT COUNT(*) FROM stress_columnstore")
                    columnstore_count = cur.fetchone()[0]
                    
                    with self.lock:
                        self.stats['consistency_checks'] += 1
                        if regular_count != columnstore_count:
                            self.stats['inconsistencies'] += 1
                            logging.warning(f"Inconsistency detected: regular={regular_count}, columnstore={columnstore_count}")
                
                time.sleep(1.0)
                
        except Exception as e:
            with self.lock:
                self.errors.append(f"Consistency check error: {e}")
        finally:
            conn.close()
    
    def run_stress_test(self, duration=30, batch_workers=2, point_workers=3, query_workers=2):
        logging.info(f"Starting stress test for {duration} seconds")
        logging.info(f"Workers: {batch_workers} batch, {point_workers} point, {query_workers} query")
        
        self.setup_tables()
        self.running = True
        
        threads = []
        
        for _ in range(batch_workers):
            t = threading.Thread(target=self.batch_insert_worker)
            threads.append(t)
            t.start()
        
        for _ in range(point_workers):
            t = threading.Thread(target=self.point_operations_worker)
            threads.append(t)
            t.start()
        
        for _ in range(query_workers):
            t = threading.Thread(target=self.query_worker)
            threads.append(t)
            t.start()
        
        consistency_thread = threading.Thread(target=self.consistency_checker)
        threads.append(consistency_thread)
        consistency_thread.start()
        
        start_time = time.time()
        try:
            while time.time() - start_time < duration:
                time.sleep(1)
                elapsed = int(time.time() - start_time)
                if elapsed % 10 == 0:
                    with self.lock:
                        logging.info(f"Progress: {elapsed}s - Stats: {self.stats}")
        
        except KeyboardInterrupt:
            logging.info("Interrupted by user")
        
        finally:
            self.running = False
            
            for t in threads:
                t.join(timeout=5)
            
            self.print_results()
            self.cleanup_tables()
    
    def print_results(self):
        print("\n" + "="*50)
        print("STRESS TEST RESULTS")
        print("="*50)
        print(f"Batch inserts: {self.stats['batch_inserts']}")
        print(f"Point inserts: {self.stats['point_inserts']}")
        print(f"Updates: {self.stats['updates']}")
        print(f"Deletes: {self.stats['deletes']}")
        print(f"Queries: {self.stats['queries']}")
        print(f"Consistency checks: {self.stats['consistency_checks']}")
        print(f"Inconsistencies found: {self.stats['inconsistencies']}")
        print(f"Errors: {len(self.errors)}")
        
        if self.errors:
            print("\nERRORS:")
            for error in self.errors[-10:]:
                print(f"  - {error}")
        
        if self.stats['inconsistencies'] == 0 and len(self.errors) == 0:
            print("\n✅ STRESS TEST PASSED - No inconsistencies or errors found")
        else:
            print("\n❌ STRESS TEST FAILED - Issues detected")
        
        print("="*50)

def main():
    parser = argparse.ArgumentParser(description='pg_mooncake stress test')
    parser.add_argument('--host', default='/home/ubuntu/.pgrx', help='PostgreSQL host (use /home/ubuntu/.pgrx for Unix socket)')
    parser.add_argument('--port', type=int, default=28817, help='PostgreSQL port (pgrx default: 28817)')
    parser.add_argument('--database', default='pg_mooncake', help='Database name')
    parser.add_argument('--user', default='ubuntu', help='Database user')
    parser.add_argument('--password', default='', help='Database password')
    parser.add_argument('--duration', type=int, default=30, help='Test duration in seconds')
    parser.add_argument('--batch-workers', type=int, default=2, help='Number of batch insert workers')
    parser.add_argument('--point-workers', type=int, default=3, help='Number of point operation workers')
    parser.add_argument('--query-workers', type=int, default=2, help='Number of query workers')
    parser.add_argument('--verbose', action='store_true', help='Enable verbose logging')
    
    args = parser.parse_args()
    
    logging.basicConfig(
        level=logging.INFO if args.verbose else logging.WARNING,
        format='%(asctime)s - %(levelname)s - %(message)s'
    )
    
    test = StressTest(
        host=args.host,
        port=args.port,
        database=args.database,
        user=args.user,
        password=args.password
    )
    
    try:
        test.run_stress_test(
            duration=args.duration,
            batch_workers=args.batch_workers,
            point_workers=args.point_workers,
            query_workers=args.query_workers
        )
    except Exception as e:
        logging.error(f"Test failed: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()

File: tests/stress_test.rs
use tokio_postgres::{Config, NoTls, types::ToSql};
use rand::Rng;
use rand::rngs::StdRng;
use rand::SeedableRng;
use std::time::{Instant, Duration};
use tokio::task::JoinHandle;
const TARGET_DURATION: Duration = Duration::from_secs(30);
const MIN_ROWS: i32 = 100000; // Minimum rows to be inserted
const MIN_TPS: f64 = 100.0; // Minimum transactions per second


#[tokio::test]
async fn stress_test() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    // Configure database connection
    let mut config = Config::new();
    config
        .dbname("pg_mooncake".to_string())
        .user("vscode".to_string())
        .password("".to_string())
        .host("localhost".to_string())
        .port(28817);

    // Connect to the database
    let (client, connection) = config.connect(NoTls).await?;
    tokio::spawn(async move {
        if let Err(e) = connection.await {
            eprintln!("connection error: {}", e);
        }
    });

    // clean up
    client
        .batch_execute("DROP TABLE IF EXISTS stress_test; 
            DROP EXTENSION IF EXISTS pg_mooncake CASCADE;
            CREATE EXTENSION IF NOT EXISTS pg_mooncake;")
        .await?;

    // Create regular table
    client
        .batch_execute(
            "CREATE TABLE stress_test (
                id INT PRIMARY KEY,
                value TEXT,
                score INT
            )",
        )
        .await?;

    // Create columnstore table
    client
        .execute(
            "CALL mooncake.create_table('stress_test_columnstore', 'stress_test')",
            &[],
        )
        .await?;
    

    // Run stress test
    let num_write_threads = 10;

    // Start concurrent writes
    let write_start_time = Instant::now();
    let mut write_handles = Vec::new();

    for thread_id in 0..num_write_threads {
        let config = config.clone();
        let handle: JoinHandle<Result<(i32, i32), Box<dyn std::error::Error + Send + Sync>>> = tokio::spawn(async move {
            let (client, connection) = config.connect(NoTls).await?;
            tokio::spawn(async move {
                if let Err(e) = connection.await {
                    eprintln!("connection error: {}", e);
                }
            });

            let mut rng = StdRng::from_entropy();
            // Two batch threads
            let batch_chance = match thread_id { 0 => 1.0, 1 => 0.1, _ => 0.0};
            let mut tx_count = 0;
            let mut batch_row_count = 0;
            
            while write_start_time.elapsed() < TARGET_DURATION {
                if rng.gen_bool(batch_chance) {
                    let batch_size = rng.gen_range(100000..500000);
                    let start_id = rng.gen_range(1..9000000);
                    client
                        .batch_execute(&format!(
                            "INSERT INTO stress_test (id, value, score)
                            SELECT 
                                i,
                                'batch_value_' || i::text,
                                (random() * 1000000)::int
                            FROM generate_series({}, {}) i
                            ON CONFLICT (id) DO UPDATE 
                            SET value = EXCLUDED.value,
                                score = EXCLUDED.score;",
                            start_id, start_id + batch_size
                        ))
                        .await?;
                    batch_row_count += batch_size;
                } else {
                    match rng.gen_range(0..3) {
                        0 => {
                            // Insert
                            let new_id = rng.gen_range(100001..200000);
                            client
                                .execute(
                                    "INSERT INTO stress_test (id, value, score) VALUES ($1, $2, $3) ON CONFLICT DO NOTHING",
                                    &[&new_id as &(dyn ToSql + Sync), &format!("new_value_{}", new_id) as &(dyn ToSql + Sync), &rng.gen_range(1..1000000) as &(dyn ToSql + Sync)],
                                )
                                .await?;
                        }
                        1 => {
                            // Update
                            let id = rng.gen_range(1..10000);
                            let new_value = format!("updated_{}", id);
                            client
                                .execute(
                                    "UPDATE stress_test SET value = $1, score = $2 WHERE id = $3",
                                    &[&new_value as &(dyn ToSql + Sync), &rng.gen_range(1..1000000) as &(dyn ToSql + Sync), &id as &(dyn ToSql + Sync)],
                                )
                                .await?;
                        }
                        2 => {
                            // Delete
                            let id = rng.gen_range(1..10000);
                            client
                                .execute(
                                    "DELETE FROM stress_test WHERE id = $1",
                                    &[&id as &(dyn ToSql + Sync)],
                                )
                                .await?;
                        }
                        _ => unreachable!(),
                    }
                    tx_count += 1;
                }
            }
            Ok((batch_row_count,tx_count))
        });
        write_handles.push(handle);
    }

    // Run analytical queries concurrently with writes using the main client
    let queries = vec![
        "SELECT COUNT(*) FROM stress_test_columnstore",
        "SELECT AVG(score) FROM stress_test_columnstore",
        "SELECT value, COUNT(*) FROM stress_test_columnstore GROUP BY value",
        "SELECT id, value, score FROM stress_test_columnstore WHERE score > 500",
        "SELECT value, AVG(score) FROM stress_test_columnstore GROUP BY value HAVING AVG(score) > 750",
    ];

    let read_start_time = Instant::now();
    let mut query_results = Vec::new();

    let mut read_id = 0;
    while read_start_time.elapsed() < TARGET_DURATION {
        let query = queries[rand::thread_rng().gen_range(0..queries.len())];
        let start_time = Instant::now();
        read_id -= 1;
        client.query("insert into stress_test (id, value, score) values ($1, 'test', 100)", &[&read_id as &(dyn ToSql + Sync)]).await?;
        let rows = client.query(query, &[]).await?;
        let duration = start_time.elapsed();
        query_results.push((query.to_string(), duration, rows.len()));
    }
    let read_duration = read_start_time.elapsed();

    let mut total_tx = 0;
    let mut total_rows = 0;
    // Collect results from all write operations
    for handle in write_handles {
        let (row_count, tx_count)= handle.await?.unwrap();
        println!("rows: {} tx: {}", row_count,tx_count);
        total_tx += tx_count;
        total_rows += row_count;
    }
    let write_duration = write_start_time.elapsed();

    let tps = total_tx as f64 / write_duration.as_secs_f64();

    // Print results
    println!("\nStress Test Results:");
    println!("Duration: {:.2?}", write_duration);
    print!("Total Rows Inserted: {}", total_rows);
    println!("Total Transactions: {}", total_tx);
    println!("Transactions per second: {:.2}", tps);
    println!("Read Queries: {}", query_results.len());
    println!("Read Duration: {:.2?}", read_duration);
    // Verify minimum throughput
    assert!(total_rows >= MIN_ROWS, "Total rows inserted {} is below minimum required {}", total_rows, MIN_ROWS);
    assert!(tps >= MIN_TPS, "Throughput {} TPS is below minimum required {} TPS", tps, MIN_TPS);

    Ok(())
} 